### PR TITLE
Add support for Inductor + symbolic shapes + training

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93059

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_functorch/aot_autograd.py
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1865]: lol  wut

File: torch/_functorch/aot_autograd.py
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1866]: LOL

File: torch/_functorch/aot_autograd.py
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:2141]: The ConstantSource here is wrong (and this matters if we actually want to perform a guard on the passed in int.) If this happened in forwards, the correct layering is to SymInt'ify the int input at the Dynamo layer. However, this is happening in backwards, which means that actually the Source here has to hook into some as yet unimplemented second level guard mechanism for the backwards function.

File: torch/_inductor/config.py
Comment (ezyang) [inline torch/_inductor/config.py:190]: 🤔

File: torch/fx/experimental/proxy_tensor.py
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:537]: don't forget to remove

File: torch/_inductor/sizevars.py
Comment (ezyang) [inline torch/_inductor/sizevars.py:462]: ahh, this was the part I was missing when I was hacking this up haha

File: torch/_inductor/sizevars.py
Comment (ezyang) [inline torch/_inductor/sizevars.py:457]: I'm a little confused here: is shape expected to be a single variable here? How do we know it always has to be a variable at this point?

Comment (ezyang) [reply torch/_inductor/sizevars.py:457]: Maybe the answer is, needed only contains variables?

**Review by ezyang — APPROVED** [review]
A good start

File: torch/_functorch/aot_autograd.py
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:1866]: This is probably what's causing all the failures :think:

Comment (ezyang) [reply torch/_functorch/aot_autograd.py:1866]: In particular, if you do it this way, I don't think you end up passing any real tensors to the backend (which means, e.g., cuda graphs doesn't work)

**Comment by ezyang** [comment]
This doesn't seem to be enough for BERT_pytorch

**Comment by ezyang** [comment]
oh whoops, gotta manually disable cuda graphs

**Comment by ezyang** [comment]
This passes BERT_pytorch. I'm doing a full sweep.

File: torch/_functorch/aot_autograd.py
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2141]: Yes, we discussed this on VC last night - ConstantSource is for something else. Let me know if you need a hand figuring out the correct source.

**Comment by voznesenskym** [comment]
LG - still interesting in how your backwards guard arch is going to look :)

**Comment by Puneet2000** [comment]
Hi, Is there an example or tutorial I can follow to understand and try out dynamic shapes better?

**Comment by soumith** [comment]
land thisssssssss!

**Review by ezyang — APPROVED** [review]

File: torch/_functorch/aot_autograd.py
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:1881]: What's the idea behind re-invoking `create_aot_dispatcher_function` now when dynamic shapes are enabled?

File: torch/_inductor/config.py
Comment (Chillee) [reply torch/_inductor/config.py]: This was just since the FX graph dumping code is broken with dynamic shapes.

File: benchmarks/dynamo/common.py
Comment (ezyang) [inline benchmarks/dynamo/common.py:1861]: This is not supposed to be necessary post https://github.com/pytorch/pytorch/pull/94469 ; can we manually prop it to the recursive aot autograd call instead?

File: torch/_inductor/sizevars.py
Comment (ezyang) [reply torch/_inductor/sizevars.py:469]: Looks like it

File: torch/_inductor/compile_fx.py
Comment (ezyang) [inline torch/_inductor/compile_fx.py:234]: This looks wrong, i is always an int so this condition will never pass

**Comment by atalman** [comment]
cc @Chillee 


This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Pull request closed without merging**
